File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const answer = [];
    let parenthesis = '';
    
    const backtracking = (open, close, str) => {
        if(str.length === n * 2){
            answer.push(str);
            return;
        }
        
        if(open < n){
            backtracking(open + 1, close, str + '(');            
        }
        
        if(close < open){
            backtracking(open, close + 1, str +')');            
        }
        
        return;
    }
    
    backtracking(0, 0, parenthesis);
    
    return answer;
};

File: Medium/0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> answer = new ArrayList();
        backtrack(answer, 0, 0, new StringBuilder(), n);
        return answer;
    }

    public void backtrack(List<String> answer, int open, int close, StringBuilder cur, int max)
    {
        if(cur.length() == max * 2){
            answer.add(cur.toString());
            return;
        }

        if(open < max){
            cur.append('(');
            backtrack(answer, open + 1, close, cur, max);
            cur.deleteCharAt(cur.length() - 1);
        }

        if(close < open){
            cur.append(')');
            backtrack(answer, open, close + 1, cur, max);
            cur.deleteCharAt(cur.length() - 1);
        }
    }
}